File: kernel/src/arch/x86_64/clocks/cmos.zig
const std = @import("std");
const root = @import("root");
const logging = root.logging;
const common = @import("../common.zig");
const port = common.port;

const logger = std.log.scoped(.x86_64_cmos);

pub const nmi_and_register_port: u16 = 0x70;
pub const data_port: u16 = 0x71;

pub const registers = struct {
    pub const seconds: u8 = 0x0;
    pub const minutes: u8 = 0x2;
    pub const hours: u8 = 0x4;
    pub const weekday: u8 = 0x6;
    pub const day_of_month: u8 = 0x7;
    pub const month: u8 = 0x8;
    pub const year: u8 = 0x9;
    pub const status_register_a: u8 = 0xA;
    pub const status_register_b: u8 = 0xB;
    pub const status_register_c: u8 = 0xC;
};

pub fn readByte(disable_nmi: bool, register: u8) u8 {
    const nmi_bit: u8 = if (disable_nmi) 0x80 else 0x00;
    port.writeByte(nmi_and_register_port, register | nmi_bit);
    return port.readByte(data_port);
}

pub fn writeByte(disable_nmi: bool, register: u8, byte: u8) void {
    const nmi_bit: u8 = if (disable_nmi) 0x80 else 0x00;
    port.writeByte(nmi_and_register_port, register | nmi_bit);
    port.writeByte(data_port, byte);
}

pub fn calibrationSleep(startTimer: fn() void) u32 {
    // Wait until next second has just started
    asm volatile (
        \\1:
        \\movb $0xA, %%al
        \\outb %%al, $0x70
        \\inb $0x71, %%al
        \\testb $0x80, %%al
        \\jz 1b
        \\2:
        \\movb $0xA, %%al
        \\outb %%al, $0x70
        \\inb $0x71, %%al
        \\testb $0x80, %%al
        \\jnz 2b
        :
        :
        : "al"
    );
    // Run measurement function
    startTimer();
    // Wait until current second has ended
    asm volatile (
        \\1:
        \\movb $0xA, %%al
        \\outb %%al, $0x70
        \\inb $0x71, %%al
        \\testb $0x80, %%al
        \\jz 1b
        \\2:
        \\movb $0xA, %%al
        \\outb %%al, $0x70
        \\inb $0x71, %%al
        \\testb $0x80, %%al
        \\jnz 2b
        :
        :
        : "al"
    );
    return 1_000_000;
}

File: kernel_rust/src/arch/x86_64/clock/cmos.rs
use super::CalibrationTimer;
use crate::arch::port;
use core::arch::asm;
use spin::Mutex;

pub mod register {
    pub const SECONDS: u8 = 0x0;
    pub const MINUTES: u8 = 0x2;
    pub const HOURS: u8 = 0x4;
    pub const WEEKDAY: u8 = 0x6;
    pub const DAY_OF_MONTH: u8 = 0x7;
    pub const MONTH: u8 = 0x8;
    pub const YEAR: u8 = 0x9;
    pub const STATUS_A: u8 = 0xA;
    pub const STATUS_B: u8 = 0xB;
    pub const STATUS_C: u8 = 0xC;
}

pub struct Cmos;

pub static CMOS: Mutex<Cmos> = Mutex::new(Cmos);

impl Cmos {
    pub unsafe fn read_byte(&self, disable_nmi: bool, register: u8) -> u8 {
        unsafe {
            let nmi_bit = (disable_nmi as u8) << 7;
            port::write_byte(port::CMOS_NMI_AND_REGISTER, register | nmi_bit);
            port::read_byte(port::CMOS_DATA)
        }
    }

    pub unsafe fn write_byte(&self, disable_nmi: bool, register: u8, byte: u8) {
        unsafe {
            let nmi_bit = (disable_nmi as u8) << 7;
            port::write_byte(port::CMOS_NMI_AND_REGISTER, register | nmi_bit);
            port::write_byte(port::CMOS_DATA, byte);
        }
    }
}

pub const CALIBRATION_TIMER: CalibrationTimer = CalibrationTimer { calibration_sleep };

unsafe fn calibration_sleep(start_timer: &mut dyn FnMut()) -> u32 {
    unsafe {
        let _cmos = CMOS.lock();
        // Wait until next second has just started
        asm!(
            "2:",
            "mov al, 0xA",
            "out 0x70, al",
            "in al, 0x71",
            "test al, 0x80",
            "jz 2b",
            "3:",
            "mov al, 0xA",
            "out 0x70, al",
            "in al, 0x71",
            "test al, 0x80",
            "jnz 3b",
            out("al") _,
            options(nomem, nostack),
        );
        // Run measurement function
        start_timer();
        // Wait until current second has ended
        asm!(
            "2:",
            "mov al, 0xA",
            "out 0x70, al",
            "in al, 0x71",
            "test al, 0x80",
            "jz 2b",
            "3:",
            "mov al, 0xA",
            "out 0x70, al",
            "in al, 0x71",
            "test al, 0x80",
            "jnz 3b",
            out("al") _,
            options(nomem, nostack),
        );
        1_000_000
    }
}
